### PR TITLE
[15482] UDPv6 support for fast-discovery-server tool and ROS_DISCOVERY_SERVER

### PR DESCRIPTION
**Pull request by MiguelBarro**
### [ROS_DISCOVERY_SERVER](https://fast-dds.docs.eprosima.com/en/latest/fastdds/env_vars/env_vars.html#ros-discovery-server)
+ Now supports IPv6 addresses for UDP.
+ The DNS names can be decorated to specify which address family to use if IPv4 and IPv6 addresses are available.

### [fast-discovery-server](https://fast-dds.docs.eprosima.com/en/latest/fastddscli/cli/cli.html#discovery)
+ Now IPv6 addresses can be specified at the cli.
+ DNS resolution supports IPv6 but available IPv4 addresses will take precedence for backward compatibility.

Testing has been updated to take into account IPv6 addresses.

Documentation https://github.com/eProsima/Fast-DDS-docs/pull/401

## Contributor Checklist
- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [x] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added. 
- [x] Any new/modified methods have been properly documented using Doxygen. 
- [x] Fast DDS test suite has been run locally. 
- [x] Changes are ABI compatible. 
- [x] Changes are API compatible. 
- [x] Documentation builds and tests pass locally. 
- [x] New feature has been added to the `versions.md` file (if applicable).
- [x] New feature has been documented/Current behavior is correctly described in the documentation. 

## Reviewer Checklist
- [ ] Check contributor checklist is correct.
- [ ] Check CI results: changes do not issue any warning.
- [ ] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

**Comment by MiguelBarro** [comment]
@richiprosima Please test this for me 🤯

**Comment by MiguelBarro** [comment]
@richiprosima Please test this for me 🤯

**Comment by MiguelBarro** [comment]
@richiprosima Please, I'm iterating, test linux 🙄

File: test/unittest/dds/participant/ParticipantTests.cpp [hunk 1]
```diff
@@ -691,7 +699,7 @@ void set_server_qos(
 
 void set_environment_variable()
 {
-    std::string environment_servers = "84.22.253.128:8888;;localhost:1234";
+    std::string environment_servers = "84.22.253.128:8888;;UDPv4:[localhost]:1234;[2a02:ec80:600:ed1a::3]:8783";
```
Comment (jsantiago-eProsima) [inline test/unittest/dds/participant/ParticipantTests.cpp:702]: I would add a check that showed that using a single DNS entry both an IPv4 and an IPv6 locators are added, along with the required UDPv6 transport.

**Review by jsantiago-eProsima — CHANGES_REQUESTED** [review]

File: src/cpp/rtps/builtin/discovery/participant/PDPClient.cpp [hunk 2]
```diff
@@ -677,95 +732,199 @@ bool load_environment_server_info(
...
 
-                            if (it->matched)
+                            // traverse submatches
+                            if (++it != mr.cend())
                             {
-                                // note stoi throws also an invalid_argument
-                                int port = stoi(it->str());
+                                std::string domain_name = it->str();
+                                std::set<std::string> ipv4, ipv6;
+                                std::tie(ipv4, ipv6) = IPLocator::resolveNameDNS(domain_name);
```
[the remark this answers is not in the document]
Comment (MiguelBarro) [reply src/cpp/rtps/builtin/discovery/participant/PDPClient.cpp:863]: Yes because is not the same.
+ The `new_locator` *stringstream* is populated with the actual environment variable text between `;`. A  `:11811` is appended and we try to read the locator using a specific method from `Locator.h`:
```
inline std::istream& operator >>( std::istream& input, Locator_t& loc)
```
+ Only if the above method fails do we try to resolve the domain name detected using the regex `ROS2_DNS_DOMAINPORT_PATTERN`. This fallback is for the case there is no specific transport prefix *UDPvxxx*. And will add all locators available.
To summarize, if a prefix is available there is only a `IPLocator::resolveNameDNS` call. And if is not too. No redundancy.
I could have done everything via regex but if we have builtin functionality we better use it (this way any builtin improvement is automatically applied to this code).

File: test/unittest/dds/participant/ParticipantTests.cpp (hunk 1 above)
Comment (MiguelBarro) [reply test/unittest/dds/participant/ParticipantTests.cpp]: And here it is: ParticipantTests.SimpleParticipantRemoteServerListConfigurationDNS

**Comment by MiguelBarro** [comment]
@richiprosima Please test this for me 🤯

**Review by jsantiago-eProsima — APPROVED** [review]
LGTM
